### PR TITLE
Implement phone input component

### PR DESCRIPTION
**Pull request by IniZio**
ref #3561

image
image

### CODE REVIEW

File: authui/package.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
     "@sentry/tracing": "7.80.0",
     "@tabler/icons": "1.107.0",
     "axios": "1.6.1",
+    "countries-list": "^3.0.6",
```
Comment (louischan-oursky) [inline authui/package.json:45]: Use libphonenumber-js and cldr-localenames-full instead. It is based on Google's libphonenumber.
https://gitlab.com/catamphetamine/libphonenumber-js
https://github.com/unicode-org/cldr-json/tree/main/cldr-json/cldr-localenames-full

In particular, please

1. `import "libphonenumber-js/max"` to ensure we are using the full data set
2. Use `getCountries()` and `getCountryCallingCode()` to generate the country list.
3. Use `cldr-localenames-full` to get the localized locale names in the **CURRENT** language. The current language is `<html lang>`
4. Sort the list according to localized locale names.

And then follow this algorithm to get the emoji flag code point https://github.com/annexare/Countries/blob/main/packages/countries/src/getEmojiFlag.ts

Example

```
String.fromCodePoint(127462 - 'A'.charCodeAt(0) + 'H'.charCodeAt(0), 127462 - 'A'.charCodeAt(0) + 'K'.charCodeAt(0)) // 🇭🇰
```

Finally we can get rid of `countries-list`.

File: authui/src/authflowv2/components/phone-input.css [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  :root {
+    --phone-input__space-x: 0.5em;
+    --phone-input__text-color: var(--color-grey-700);
+    --phone-input__font-size: var(--typography-body-large__font-size);
+    --phone-input__line-height: var(--typography-body-large__line-height);
+    --phone-input__letter-spacing: var(--typography-body-large__letter-spacing);
+    --phone-input__font-weight: var(--typography-body-large__font-weight);
+    --phone-input__font-family: twemoji-countries, Inter;
+
+    --phone-input__trigger__ring-color: var(--color-grey-100);
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/phone-input.css:11]: We follow BEM and BEM does not have two elements. So it should be named as `--phone-input__trigger-ring-color`.

File: authui/src/authflowv2/components/phone-input.css [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+  .phone-input__options {
+    @apply flex;
+    @apply flex-col;
+    @apply p-4;
+
+    max-height: 288px;
+    overflow-y: auto;
+
+    @apply space-y-[var(--phone-input\_\_options\_\_space-y)];
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/phone-input.css]: use gap.

File: authui/src/customSelect.ts [hunk 4]
```diff
@@ -1,38 +1,160 @@
...
+    }).then(({ x, y }) => {
+      Object.assign(this.dropdownTarget.style, {
+        left: `${x}px`,
+        top: `${y}px`,
+      });
+    });
+
+    this.triggerTarget.addEventListener("keydown", this.handleKeyDown);
+    this.dropdownTarget.addEventListener("keydown", this.handleKeyDown);
+    document.addEventListener("click", this.handleClickOutside);
```
Comment (louischan-oursky) [inline authui/src/customSelect.ts:71]: Have you tried https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog ? Does it work?

File: authui/src/authflowv2/components/phone-input.css [hunk 5]
```diff
@@ -103,6 +103,42 @@
...
+  .phone-input__search {
+    @apply relative;
+    @apply flex;
+    @apply flex-row;
+    @apply items-center;
+
+    border-bottom: 1px solid var(--color-grey-100);
+  }
+
+  .phone-input__search-input {
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/phone-input.css]: Can we reuse the input field component here?

File: authui/src/authflowv2/components/phone-input.css [hunk 6]
```diff
@@ -20,30 +20,11 @@
...
-    --phone-input__options-height: 100%;
-    --phone-input__options__space-y: 0.5em;
-
-    --phone-input__option__border-radius: calc(
-        var(--phone-input__font-size) * 0.75
-      )
-      /* 12 / 16 ratio to font size */;
-    --phone-input__option__background-color--active: var(--color-grey-100);
-
+    --phone-input__error-message-margin-top: 0.5em;
```
Comment (louischan-oursky) [inline authui/src/authflowv2/components/phone-input.css:27]: `error-message-spacing`

File: authui/src/customSelect.ts [hunk 7]
```diff
@@ -50,6 +50,7 @@ export class CustomSelectController extends Controller {
 
   keyword = "";
   value?: string;
+  focusedValue?: string;
```
Comment (louischan-oursky) [inline authui/src/customSelect.ts:53]: Do not introduce redundant instance properties when the source of truth is the attributes of some elements in our control.

**Review by louischan-oursky — COMMENTED** [review]
Just left some preliminary comments.

File: authui/src/customSelect.ts (hunk 4 above)
Comment (IniZio) [reply authui/src/customSelect.ts:71]: Couldn't get focus to work when using dialog, will try again after addressing other comments.

File: authui/src/authflowv2/components/phone-input.css (hunk 5 above)
Comment (IniZio) [reply authui/src/authflowv2/components/phone-input.css]: The search input's style is pretty different from the input component, or you meant adding a variant to input?

File: authui/src/customSelect.ts (hunk 4 above)
Comment (IniZio) [reply authui/src/customSelect.ts:83]: Able to make dialog work with current country select. 

The dialog features used: `::backdrop`, `.show` should have ~95% on browser support.

**Pull request merged**